### DISCUSSION TITLE
MultiOutput KeOpsKernel

### DISCUSSION THREAD

**Discussion by arthus701** [cornellius-gp/gpytorch#1532] [Q&A]
Hi everyone,
I am struggling with a memory efficient implementation of the kernels mentioned [here](#1387) and [here](#1466). Conceptually, they are kernels for vector valued outputs, that include some linear transformation/rotation of the vector coordinate system. After the discussion over at #1453, I am trying to implement them as KeOpsKernels.

Does anyone have experience with pulling off something similar to [this](https://github.com/cornellius-gp/gpytorch/discussions/1504#discussioncomment-409376) MultiOutput-Kernel with KeOps? Or is the current restriction to two "indexing" dimensions of KeOps permitting this? I guess you wouldn't need more symbolic dimensions, just a way to broadcast or extend the KeOps-tensors.

If it helps I could also provide the PyTorch implementation that I currently have.

Thank you,
Arthus

**Comment by gpleiss** [comment]
Hey @arthus701 - I saw you have an example that you're working through with the KeOps folks (https://github.com/getkeops/keops/issues/156). Let me know how this is going.
